File: neetcode/1. arrays/48.designHashmap.cpp
// Design a HashMap without using any built-in hash table libraries.

// Implement the MyHashMap class:

// MyHashMap() initializes the object with an empty map.
// void put(int key, int value) inserts a (key, value) pair into the HashMap. 
// If the key already exists in the map, update the corresponding value.
// int get(int key) returns the value to which the specified key is mapped, 
// or -1 if this map contains no mapping for the key.
// void remove(key) removes the key and its corresponding value if the map contains 
// the mapping for the key.

#include <limits>
#include <vector>

struct Node {
public:
  int key, val;
  Node* next;
  Node(int k, int v, Node* n) {
    key = k;
    val = v;
    next = n;
  }
};

// class with Hash
class MyHashMap {
public:
  const static int size = 19997;
  const static int mult = 12582917;
  Node* data[size] = {};
  
  int hash(int key) {
    return (int)((long)key * mult  % size);
  }

  void put(int key, int val) {
    remove(key);
    int h = hash(key);
    Node* node = new Node(key, val, data[h]);
    data[h] = node;
  }

  int get(int key) {
    int h = hash(key);
    Node* node = data[h];
    for (; node != NULL; node = node->next) {
      if (node->key == key) {
        return node->val;
      }
    }
    return -1;
  }

  void remove(int key) {
    int h = hash(key);
    Node* node = data[h];
    if (node == NULL) return;
    if (node->key == key) {
      data[h] = node->next;
      delete node;
    } else for (; node->next != NULL; node = node->next) {
      if (node->next->key == key) {
        Node* temp = node->next;
        node->next = temp->next;
        delete temp;
        return;
      }
    }
  }
};

// class with vector
class MyHashMap {
public:
  int data[1000001];
  MyHashMap() {
    std::fill(data, data + 1000000, -1);
  }
  void put(int key, int val) {
    data[key] = val;
  }
  int get(int key) {
    return data[key];
  }
  void remove(int key) {
    data[key] = -1;
  }
};
/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */

File: neetcode/1. arrays/48.designHashmap.py
# Design a HashMap without using any built-in hash table libraries.

# Implement the MyHashMap class:

# MyHashMap() initializes the object with an empty map.
# void put(int key, int value) inserts a (key, value) pair into the HashMap. 
# If the key already exists in the map, update the corresponding value.
# int get(int key) returns the value to which the specified key is mapped, 
# or -1 if this map contains no mapping for the key.
# void remove(key) removes the key and its corresponding value if the map contains 
# the mapping for the key.

class ListNode:
    def __init__(self, key=-1, val=-1, next=None):
        self.key = key
        self.val = val
        self.next = next

class MyHashMap:
    def __init__(self):
        self.map = [ListNode() for i in range(1000)]

    def hashcode(self, key):
        return key % len(self.map)
    
    def put(self, key: int, value: int) -> None:
        cur = self.map[self.hashcode(key)]
        while cur.next:
            if cur.next.key == key:
                cur.next.val = value
                return
            cur = cur.next
        cur.next = ListNode(key, value)
    
    def get(self, key: int) -> int:
        cur = self.map[self.hashcode(key)].next
        while cur and cur.key != key:
            cur = cur.next
        if cur:
            return cur.val
        return -1
    
    def remove(self, key: int) -> None:
        cur = self.map[self.hashcode(key)]
        while cur.next and cur.next.key != key:
            cur = cur.next
        if cur and cur.next:
            cur.next = cur.next.next


# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)